src/Autonomous.cpp

```diff
@@ -243,23 +264,99 @@ void Autonomous::autonomyIter()
+			}
+			// we have already seen the first post
+			else
+			{
+				point_t post_1 = landmarkFilter.get();
+				point_t center = (post_1 + post_2) / 2;
+
+				// Use the center as the drive target to avoid being close to one post but far from the other
+				driveTarget.topRows(2) = center.topRows(2);
+
+				// True if we are close enough to have sufficiently accurate landmark measurements
+				bool inRange = std::isnan(gate_targets.first(2)) && dist(pose, center, 0.0) < 8.0;
+				// True if we are within refining distance
+				bool shouldRefine = !std::isnan(gate_targets.first(2)) && std::isnan(gate_targets.second(2)) && dist(pose, gate_targets.first, 0.0) < 1.5;
+
+				// we have not set gate targets yet or have potentially inaccurate gate landmark measurements
+				if (inRange || shouldRefine)
+				{
+					point_t offset{(post_1(1) - post_2(1)) / 2, (post_2(0) - post_1(0)) / 2, 0};
+					pose_t goal_1 = center + 2 * offset;
+					pose_t goal_2 = center - 2 * offset;
+
+					// Choose the closest goal to be the first target
+					if (dist(pose, goal_1, 0) < dist(pose, goal_2, 0))
+					{
+						gate_targets = std::make_pair(goal_1, goal_2);
+					}
+					else
+					{
+						gate_targets = std::make_pair(goal_2, goal_1);
+					}
+
+					// Angle between targets
+					double angle = std::atan2(gate_targets.second(1) - gate_targets.first(1), gate_targets.second(0) - gate_targets.first(0));
+					// Angle between current post and first target
+					gate_targets.first(2) = std::atan2(gate_targets.first(1) - pose(1), gate_targets.first(0) - pose(0));
+					// Mark targets as unrefined if not refining
+					gate_targets.second(2) = shouldRefine ? angle : NAN;
+
+					// Replan
+					plan_cost = INFINITE_COST;
+
+					// Align to target to improve planning
+					state = NavState::GATE_ALIGN;
+				}
+
+				// we have already seen the first post and have set gate targets
+				if (!std::isnan(gate_targets.first(2)))
+				{
+					// Update gate targets
+					if (!std::isinf(gate_targets.first(2)) &&
+						dist(pose, gate_targets.first, 0.0) < 0.2)
+					{
+						// Arrived at first gate target, mark as complete
+						gate_targets.first = pose_t(INFINITY, INFINITY, INFINITY);
+
+						// Update target angle of second target to account for inaccuracies
+						gate_targets.second(2) = std::atan2(gate_targets.second(1) - pose(1), gate_targets.second(0) - pose(0));
+
+						// Turn to desired angle
+						state = NavState::GATE_ALIGN;
+					}
+					if (std::isinf(gate_targets.first(2)) && dist(pose, gate_targets.second, 0.0) < 0.3)
+					{
+						// Arrived at second gate target
+						gate_targets.second = pose_t(INFINITY, INFINITY, INFINITY);
+
+						// No further action needed since goal has been reached
+						already_arrived = true;
+						return;
+					}
+					driveTarget = std::isinf(gate_targets.first(2)) ? gate_targets.second : gate_targets.first;
+					// Avoids weird planning behavior when close to the second target
```

Comment:
This is a good catch. As mentioned above, we should re-plan whenever we reach a key waypoint (such as the first gate target).